tls_main.c

```diff
@@ -567,47 +588,70 @@ static int do_tls_setsockopt_conf(struct sock *sk, char __user *optval,
 	return rc;
 }
 
-#define TLS_LUA 99
-#define lua_error(msg) pr_warn("[lua] %s - %s\n", __func__, msg);
-
-static int do_tls_loadlua(char __user *optval, unsigned int optlen)
+static int do_tls_lua_load(struct sock *sk, char __user *optval,
+			   unsigned int optlen)
 {
+	struct tls_context *ctx = tls_get_ctx(sk);
+	char *script;
 	int rc = 0;
-	char *file;
-	lua_State *L;
 
-	file = kmalloc(optlen, GFP_KERNEL);
-	if (!file) {
-		lua_error("no memory");
+	script = kmalloc(optlen, GFP_KERNEL);
+	if (!script) {
+		TLS_LUA_ERROR("no memory");
 		return -ENOMEM;
 	}
-	rc = copy_from_user(file, optval, optlen);
+
+	rc = copy_from_user(script, optval, optlen);
 	if (rc) {
 		rc = -EFAULT;
-		goto end;
+		goto out;
 	}
 
-	L = luaL_newstate();
-	if (L == NULL) {
-		lua_error("no memory");
-		return -ENOMEM;
+	if (!ctx->L) {
+		ctx->L = luaL_newstate();
+		if (!ctx->L) {
+			TLS_LUA_ERROR("no memory");
+			rc = -ENOMEM;
+			goto out;
+		}
+		luaL_openlibs(ctx->L);
 	}
-	luaL_openlibs(L);
 
-	if (luaL_dostring(L, file)) {
-		lua_error(lua_tostring(L, -1));
+	if (luaL_dostring(ctx->L, script)) {
+		TLS_LUA_ERROR(lua_tostring(ctx->L, -1));
 		rc = -ECANCELED;
-		goto end;
+		goto out;
 	}
 
-end:
-	lua_close(L);
-	kfree(file);
+out:
+	kfree(script);
 	return rc;
 }
 
-static int do_tls_setsockopt(struct sock *sk, int optname,
-			     char __user *optval, unsigned int optlen)
+static int do_tls_lua_setrecv(struct sock *sk, char __user *optval,
+			      unsigned int optlen)
+{
+	struct tls_context *ctx = tls_get_ctx(sk);
+	int rc = 0;
+
+	kfree(ctx->recv_entry);
+	ctx->recv_entry = kmalloc(optlen, GFP_KERNEL);
```

Comment:
Maybe we should create a fixed size array for the function name, instead of just using a pointer. @tammela what do you think?

Reply:
It looks better indeed

Reply:
@MXWXZ Please address this comment as well

Reply:
maybe 256 is enough?

Reply:
sounds good